Use checkbox menu in crate selection.

poelzi opened the pull request:
Switch from normal QAction menu to a composition of QCheckBox and QActionWidget.
The checkbox shows in which crates the selection is in.
Changing the crates selection does not collapse the menu, which allows
much easier categorization of tracks without going through the menu from scratch.

daschuer commented:
Thank you for this PR. Nice Idea.

It work well for a single track, unfortunately not well for multiple tracks. I think we need a three state check box or something, for the case where some of the selected tracks are in a crate and some not. 

Nit: Please move the pointer asterisk next to QWidget to meet our coding style.  

Be fore merge, we need your permission.
Please sign https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ and comment here when done.

@gramanas: how does this merge with your crate hierarchy branch?

gramanas commented:
I can't check it atm, but this is based on master and not the new library redesign. Since it only changes widget files, I don't think it'll be hard to resolve the merge conflicts, but I can definitely see some.

poelzi commented:
I implemented the tristate when multiple tracks are selected that do not share a crate fully.
Fixed your styling comment as well (I hope ;))

daschuer commented on src/widget/wtracktableview.cpp:
```diff
@@ -828,20 +831,45 @@ void WTrackTableView::contextMenuEvent(QContextMenuEvent* event) {
+            // FIXME: we could use the tristate feature here and show grey in case of multi track selection
+            // in which not all track are in the crate
+            if(currentCrates.contains(crate.getId())) {
+                if(currentCrates[crate.getId()].size() != trackIds.size()) {
```
Ok, this works, but it took me quite a while to find out how. 
I think this requires some comments.
currentCrates should also be renamed ... mmm 
or can we move the size compare if condition to collectCrateTrackListOfTracks ?

poelzi replied:
My thought was, that the API maybe more useful in the future. Giving a list of tracks and getting a hash with the crates and the containing tracks. Maybe I was overthinking it. I changed the api to just count the tracks in the crate.

daschuer commented on src/widget/wtracktableview.cpp:
```diff
@@ -828,20 +831,45 @@ void WTrackTableView::contextMenuEvent(QContextMenuEvent* event) {
+
+            // FIXME: we could use the tristate feature here and show grey in case of multi track selection
+            // in which not all track are in the crate
+            if(currentCrates.contains(crate.getId())) {
```
contains() and a subsequent [] is a double hash lookup. Use find();

daschuer commented on src/library/crate/cratestorage.cpp:
```diff
@@ -507,6 +507,29 @@ QSet<CrateId> CrateStorage::collectCrateIdsOfTracks(const QList<TrackId>& trackI
+        while (crateTracks.next()) {
+            DEBUG_ASSERT(crateTracks.trackId() == trackId);
+            CrateId cid = crateTracks.crateId();
+            if(!trackCrates.contains(cid)) {
```
this if condition is redundant, because operator[] has already such a code.

daschuer reviewed (changes requested):
Thank you for the changes. 
It works nice now.

I have added some comments most of them are to improve the readability of the code.

poelzi commented on src/widget/wtracktableview.cpp:
```diff
@@ -1466,22 +1494,51 @@ void WTrackTableView::addSelectionToPlaylist(int iPlaylistId) {
-        crateId = CrateFeatureHelper(
-                m_pTrackCollection, m_pConfig).createEmptyCrate();
+    // we need to disable tristate again as the mixed state will now be gone and can't be brought back
+    if (qc->property("tristate").toBool() == true) {
```
I tried different techniques for hours, but this was the only only I got working properly. I'm quite open if someone gets this running using more slick code, but as I think it's none critical code from performance aspects, I think it's OK.
I will write some more optimal SQL queries in a separate pull request which should be a much better optimization.

daschuer replied:
See https://github.com/poelzi/mixxx/pull/1

poelzi commented:
Thanks for the review. I haven't user QT C++ before and haven't touched C++ for a very long time.
I hope I have addressed all points. Unfortunately I haven't found a way to avoid string properties in a slick way, but I don't think it's critical in any way.

uklotzde commented on src/library/crate/cratestorage.cpp:
```diff
@@ -507,6 +507,24 @@ QSet<CrateId> CrateStorage::collectCrateIdsOfTracks(const QList<TrackId>& trackI
+        // the amount of code and the number of prepared SQL queries.
+        CrateTrackSelectResult trackCrates(selectTrackCratesSorted(trackId));
+        while (trackCrates.next()) {
+            rv[trackCrates.crateId()] += 1;
```
Are you sure that the integer value in each entry is explicitly initialized with 0??? The fundamental type _int_ does not have a default constructor and QHash will leave it uninitialized when inserting a new key/value pair.

poelzi replied:
As far as I understood the QList documentation. It states that such types are initialized with 0.

uklotzde commented:
Very nice, I like this feature 👍 Now I can delete my custom branch that allowed me to delete selected tracks from a crate. But I omitted to add visual feedback, if those tracks are actually contained in crates. That's why I never created a PR.

uklotzde commented:
Proposal: The name of the menu entry should be changed from "Add to Crate" to just "Crates"!

uklotzde commented on src/widget/wtracktableview.cpp:
```diff
@@ -828,20 +831,49 @@ void WTrackTableView::contextMenuEvent(QContextMenuEvent* event) {
+
+        const QList<TrackId> trackIds = getSelectedTrackIds();
+
+        QHash<CrateId, int> currentCrates(m_pTrackCollection->crates().\
```
Unfortunately many queries are executed even if I don't open the _Crates_ menu entry. Each right click on selected tracks now has a noticeable delay when you have many crates. The list of crates with their state should only be populated when the corresponding sub-menu entry becomes visible.

poelzi replied:
Ok, I will look into optimizing it and change the name to crates :)

daschuer commented on src/widget/wtracktableview.cpp:
```diff
@@ -828,20 +831,49 @@ void WTrackTableView::contextMenuEvent(QContextMenuEvent* event) {
         Crate crate;
         while (allCrates.populateNext(&crate)) {
-            auto pAction = std::make_unique<QAction>(crate.getName(), m_pCrateMenu);
+            auto pAction = std::make_unique<QWidgetAction>(m_pCrateMenu);
```
we have recently introduced "make_parented" to document the qt object tree ownership and to avoid the unused unique_ptr nature.

uklotzde commented:
@poelzi Any plans to address the open comments?

I must admit that this is a really helpful feature that I merge into my private builds. I would like to see it in the next release(s). I usually have around 50+ crates and the performance penalties seem to be acceptable. We could even discuss to deliver the performance improvements in a follow-up release if (hopefully) we are able to return to a more frequent release cycle ;)

poelzi commented:
I will hopefully get back addressing all the comments I have on my open pull requests and this one is annoying me as well ;) I had no working development machine for a while or was quite busy.

poelzi commented:
I was able to optimize everything into one query with a subselect and reuse the CratesSummery code.
Got even slicker :)

poelzi committed:
Use checkbox menu in crate selection.

Switch from normal QAction menu to a composition of QCheckBox and QActionWidget.
The checkbox shows in which crates the selection is in.
Changing the crates selection does not collapse the menu, which allows
much easier categorization of tracks without going through the menu from scratch.

Use tristate when multiple tracks are selected.

When mulitple tracks are selected which do not share a crate, use
the tristate partially selected to indicate this.

Fix styling of pointers.

Use QVariant to transport CrateId in checkbox

Use optimized SQL query to select crates and count tracks.

Unfortunatelly QSqlQuery has no way of binding QLists
in WHERE x IN statements.

Use make_partented as suggested by daschuer

rename "Add to Crates" -> "Crates" to reflect function better

uklotzde committed:
Optimize formatting of string lists for SQL queries

poelzi committed:
Merge pull request #2 from uklotzde/feature/crate_selectbox

Optimize formatting of string lists for SQL queries

Be-ing commented:
@poelzi for the future, it's easier to review changes since the last review if new commits are added instead of rebasing.

Be-ing commented on src/library/crate/cratestorage.cpp:
```diff
@@ -468,6 +468,42 @@ CrateTrackSelectResult CrateStorage::selectTrackCratesSorted(TrackId trackId) co
 }
 
+CrateSummarySelectResult CrateStorage::selectCratesWithTrackCount(const QList<TrackId>& trackIds) const {
+
```
nitpick: extra blank line here

Be-ing commented on src/library/crate/cratestorage.cpp:
```diff
@@ -468,6 +468,42 @@ CrateTrackSelectResult CrateStorage::selectTrackCratesSorted(TrackId trackId) co
+    // Using bindValue did not work, only constructing the SQL string befor.
+    // As TrackId is a int, we are safe here
+    QStringList idstrings;
+    foreach(TrackId id, trackIds) {
```
Use standard C++11 `for` loop instead of old Qt `foreach` macro.

Be-ing commented on src/library/crate/cratestorage.cpp:
```diff
@@ -468,6 +468,42 @@ CrateTrackSelectResult CrateStorage::selectTrackCratesSorted(TrackId trackId) co
+    } else {
+        return CrateSummarySelectResult();
+    }
+
```
extra blank line

Be-ing reviewed (commented):
just some small style nitpicks

poelzi commented:
Be-ing: I think your nit picks where fixed by the optimization Uwe Klotz did.

Be-ing commented:
Oh, you're correct. Sorry for the noise.

uklotzde commented:
The idea of lazily populating the context sub-menus is also mentioned in this bug report:
https://bugs.launchpad.net/mixxx/+bug/1395022
Executing even more synchronous database queries to obtain results that might never be needed will impact the user experience even more.

Again, I like the functionality and would accept the lag in the first place. But we should think about how this can be fixed ASAP.

poelzi commented:
@uklotzde i totally agree, but the current iteration does not make any more SQL calls  then before.

`sqlite> SELECT *, (    SELECT COUNT(*) FROM crate_tracks WHERE crates.id = crate_tracks.crate_id and crate_tracks.track_id in (29, 31, 28, 19, 15,135, 267) ) AS track_count, 0 as track_duration FROM crates ORDER BY name;
6|Haha1|0|1|0|0|3|0
4|blbl|0|1|0|0|0|0
3|blubb|0|1|0|0|0|0
5|huhu|0|1|0|0|3|0
10|huhuihaenuae|0|1|0|0|0|0
1|test1|0|1|1|0|2|0
2|test2|0|1|0|0|1|0
7|turaine|0|1|0|0|2|0
8|uaierutaors|0|1|0|0|0|0
9|uiaeuaie|0|1|0|0|1|0
-------- scanstats --------
Loop  1: SCAN TABLE crates USING INDEX sqlite_autoindex_crates_1
         nLoop=1        nRow=10       estRow=1048576  estRow/Loop=1.04858e+06
-------- subquery 1 -------
Loop  1: SEARCH TABLE crate_tracks USING COVERING INDEX sqlite_autoindex_crate_tracks_1 (crate_id=? AND track_id=?)
         nLoop=10       nRow=12       estRow=80       estRow/Loop=8       
---------------------------
Run Time: real 0.000 user 0.000000 sys 0.000000
sqlite> 
`
It's 2 loops instead of one but fully covered by indexes.

uklotzde commented:
The lazy initialization of 2nd level actions affects multiple use cases and is a cross-cutting concern and could be solved in a separate task.

I vote for integrating this extremly useful feature before improving the performance of **all** context menus that require database access.

poelzi committed:
Fix Dataloss in history view while changing crate selection

If the song changes while the current history view is open, the selection of
songs is lost. Therefor all crate changes are lost.

The history view saves and restores the selection through a new usefull api for selecting tracks in the wtracklistview.

Be-ing commented:
I test this and it works nicely. It would be cool to have this for Playlists too, but before that is implemented I think this needs to be lazy initialized or right clicking tracks will become even more terribly slow.

uklotzde commented:
LGTM.

Some nitpicking about typos and coding style. But I don't think we need another round here. I would like to get the approval of 1 or 2 other reviewers before merging anything myself.

@daschuer How do we verify that the author has signed the contributor agreement?

daschuer commented:
@poelzi is already a contributor.
Do you have a Launchpad account? Than I can add you to the Mixxx Contributors Group.

poelzi commented:
@daschuer it's poelzi  on launchpad :)

poelzi committed:
Merge branch 'master' of https://github.com/mixxxdj/mixxx into feature/crate_selectbox

uklotzde commented:
@poelzi Do you have time to address Be's comments and resolve the pending merge conflict, Daniel? I would like to have this in 2.1, because it has proven to be a valuable feature 👍

poelzi committed:
Merge branch 'master' of https://github.com/mixxxdj/mixxx into feature/crate_selectbox

poelzi committed:
style fixes according to be's comments

Be-ing commented:
LGTM, thank you! @poelzi would you be interested in implementing the lazy initialization to alleviate https://bugs.launchpad.net/mixxx/+bug/1733200 for 2.1?

poelzi commented:
@Be-ing kinda added the required infrastructure in another branch already. I added a widget (qlistview) that displays the crates of the selected track. I used a background thread to fetch the information in background and update the gui when done. If we decide to use one thread for sql queryies and gui support or one thread per component. https://ibb.co/dJsuf6

Be-ing commented:
@uklotzde ready for merge?

uklotzde commented on src/widget/wtracktableview.cpp:
```diff
@@ -1373,6 +1401,31 @@ QList<TrackId> WTrackTableView::getSelectedTrackIds() const {
     return trackIds;
 }
 
+void WTrackTableView::setSelectedTracks(QList<TrackId> trackIds) {
```
This parameter should be passed as const-ref to avoid a deep copy of the list. Unfortunately the range-based for loops do not work well with Qt containers. Declaring the loop variable as const-ref is not enough. The container itself must be const! There is already a link in our [coding guidelines](https://www.mixxx.org/wiki/doku.php/coding_guidelines#range-based_for_loops).

We have this issue at many places and I have done this wrong until recently. Checking the code with [clazy](https://github.com/KDE/clazy) could reveal those hidden performance killers.

daschuer replied:
I will fix this after merge

Be-ing replied:
I already did in #1427

daschuer replied:
included in 187ec3d5fa231f2ee1259fe724ca50386a8ff9fc by @Be-ing

uklotzde commented:
Almost ready. Just one minor issue that we should fix, now that we know better ;)

Be-ing commented:
Thank you for this big usability improvements for crates! Let's continue this momentum to make Mixxx 2.2 the best music library management software. :)
